### PR TITLE
Message export should include delivery report status.

### PR DESCRIPTION
**Pull request by smn**
Useful to know for analysis of SMS based conversations.

### CODE REVIEW

**Comment by smn** [comment]
Merging #851 into this to continue work started there. That will need to be reviewed & land first.

**Comment by smn** [comment]
Also merging in #853

File: go/conversation/tasks.py [hunk 1]
```diff
@@ -95,22 +101,41 @@ def export_conversation_messages_unsorted(account_key, conversation_key):
...
+                            reverse=True)
+            row = dict((field, unicode(message.payload[field]))
+                       for field in conversation_export_field_names
+                       if field in message)
+            row['direction'] = 'outbound'
+            delivery_reports = [event for event in events
+                                if event['event_type'] == 'delivery_report']
+            row['delivery_status'] = get_delivery_status(delivery_reports)
+            network_events = [event for event in events
+                              if event['event_type'] in ['ack', 'nack']]
```
Comment (hodgestar) [inline go/conversation/tasks.py]: I think we should split the bodies of these loops out into functions, just so this method is shorter and we're sure we're not accidentally re-using local variables we don't want to? Maybe `row_for_outbound_messsage(...)` and `row_for_inbound_message(...)`?

**Comment by smn** [comment]
doh I need to land the changes I made to Vumi.

File: go/base/utils.py [hunk 2]
```diff
@@ -124,6 +124,41 @@ def writerows(self, rows):
             self.writerow(row)
 
 
+class UnicodeDictWriter(object):
+    """
+    Essentially a copy of the built in csv.DictWriter with the exception
+    that the ``extrasaction`` keyword argument has been dropped.
+    This class will always raise an exception if the keys in the dict
+    do not match the known fieldnames.
+    """
```
Comment (hodgestar) [inline go/base/utils.py:139]: I think the docstring should mention that it builds on the `UnicodeCSVWriter` since that's the classes reason for existing?

**Comment by smn** [comment]
Travis will be unhappy about this PR until praekelt/vumi#720 lands.

**Comment by hodgestar** [comment]
Looks good. Left two minor comments.

**Comment by smn** [comment]
@hodgestar thanks, ready for re-review.

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**